### ISSUE TITLE
Parameter normalization issue

### ISSUE THREAD

**Issue by Shane32** [linq2db/linq2db#4228]
### Describe your issue

After upgrading from linq2db 5.1.1 to 5.2.2 last week, I am receiving intermittent exceptions in production - see below.  The query is a union of several basic searches; each part searches a certain string field with StartsWith, returning the row id (an int64), and a constant (an int32) representing the table searched.  The query works fine in development (even when connected to the production database) and is intermittent in production, even with the exact same inputs.  I am unable to replicate the failure within tests.  When running the query locally and grabbing a copy of the generated SQL, it runs successfully against the production database and correctly returns a single row of data.

```
Exception message: Index and length must refer to a location within the string. (Parameter 'length')
Stack trace:
   at System.String.Substring(Int32 startIndex, Int32 length)
   at LinqToDB.DataProvider.UniqueParametersNormalizer.Normalize(String originalName)
   at LinqToDB.SqlProvider.OptimizationContext.AddParameter(SqlParameter parameter)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildExpression(ISqlExpression expr, Boolean buildTableName, Boolean checkParentheses, String alias, Boolean& addAlias, Boolean throwExceptionIfTableNotFound)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildLikePredicate(Like predicate)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildPredicate(ISqlPredicate predicate)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildPredicate(Int32 parentPrecedence, Int32 precedence, ISqlPredicate predicate)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildSearchCondition(SqlSearchCondition condition, Boolean wrapCondition)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildSearchCondition(Int32 parentPrecedence, SqlSearchCondition condition, Boolean wrapCondition)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildWhereClause(SelectQuery selectQuery)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildSelectQuery(SqlSelectStatement selectStatement)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildSql(Int32 commandNumber, SqlStatement statement, StringBuilder sb, OptimizationContext optimizationContext, Int32 indent, Boolean skipAlias)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildSql(Int32 commandNumber, SqlStatement statement, StringBuilder sb, OptimizationContext optimizationContext, Int32 indent, Boolean skipAlias)
   at LinqToDB.SqlProvider.BasicSqlBuilder.BuildSql(Int32 commandNumber, SqlStatement statement, StringBuilder sb, OptimizationContext optimizationContext, Int32 startIndent)
   at LinqToDB.Data.DataConnection.QueryRunner.GetCommand(DataConnection dataConnection, IQueryContext query, IReadOnlyParameterValues parameterValues, Boolean forGetSqlText, Int32 startIndent)
   at LinqToDB.Data.DataConnection.QueryRunner.CreateExecutionQuery(DataConnection dataConnection, IQueryContext context, IReadOnlyParameterValues parameterValues, Boolean forGetSqlText)
   at LinqToDB.Data.DataConnection.QueryRunner.SetQuery(IReadOnlyParameterValues parameterValues, Boolean forGetSqlText)
   at LinqToDB.Data.DataConnection.QueryRunner.ExecuteReaderAsync(CancellationToken cancellationToken)
   at LinqToDB.Linq.QueryRunner.ExecuteQueryAsync[T](Query query, IDataContext dataContext, Mapper`1 mapper, Expression expression, Object[] ps, Object[] preambles, Int32 queryNumber, Func`2 func, TakeSkipDelegate skipAction, TakeSkipDelegate takeAction, CancellationToken cancellationToken)
   at LinqToDB.Linq.QueryRunner.ExecuteQueryAsync[T](Query query, IDataContext dataContext, Mapper`1 mapper, Expression expression, Object[] ps, Object[] preambles, Int32 queryNumber, Func`2 func, TakeSkipDelegate skipAction, TakeSkipDelegate takeAction, CancellationToken cancellationToken)
   at LinqToDB.Linq.ExpressionQuery`1.GetForEachAsync(Action`1 action, CancellationToken cancellationToken)
   at LinqToDB.Linq.ExpressionQuery`1.GetForEachAsync(Action`1 action, CancellationToken cancellationToken)
   at LinqToDB.AsyncExtensions.ToListAsync[TSource](IQueryable`1 source, CancellationToken token)
   at HiveGraphQL.Query.<>c.<<ConfigureBarcodeField>b__15_0>d.MoveNext() in /home/runner/work/HiveServer/HiveServer/HiveGraphQL/_Query.cs:line 1464
--- End of stack trace from previous location ---
   at GraphQL.Execution.ExecutionStrategy.ExecuteNodeAsync(ExecutionContext context, ExecutionNode node) in /_/src/GraphQL/Execution/ExecutionStrategy.cs:line 502
```

The `UniqueParametersNormalizer` class was added in #4146 so I would guess that it's related to the added code.

Please let me know what I should try next.  I will plan to try downgrading to 5.2.1 next week and see if that appears to clear up the issue.

### Code

Here is a code sample with actual table/column names removed by ChatGPT.

```cs
var queries = new List<IQueryable<GenericSearchResultModel>>();

queries.Add(
    from table1 in db.Table1
    join table2 in db.Table2 on table1.Id equals table2.Table1Id
    where table1.Deleted == false && table2.Field1.StartsWith(search)
    select new GenericSearchResultModel { Id = table1.Id, Type = GenericModelType.Type1 });
queries.Add(
    from table3 in db.Table3
    where table3.Field2!.StartsWith(search)
    select new GenericSearchResultModel { Id = table3.Id, Type = GenericModelType.Type2 });
queries.Add(
    from table4 in db.Table4
    where table4.Field2.StartsWith(search)
    select new GenericSearchResultModel { Id = table4.Table3Id, Type = GenericModelType.Type3 });
queries.Add(
    from table1 in db.Table1
    where table1.Field3.StartsWith(search) && table1.Deleted == false
    select new GenericSearchResultModel { Id = table1.Id, Type = GenericModelType.Type1 });
queries.Add(
    from table1 in db.Table1
    join table5 in db.Table5 on table1.Id equals table5.Table1Id
    where table5.Field4 == search && table1.Deleted == false
    select new GenericSearchResultModel { Id = table1.Id, Type = GenericModelType.Type1 });
queries.Add(
    from table1 in db.Table1
    where table1.Field5!.StartsWith(search) && table1.Deleted == false
    select new GenericSearchResultModel { Id = table1.Id, Type = GenericModelType.Type1 });
queries.Add(
    from table6 in db.Table6
    where table6.Name.StartsWith(search) && table6.Deleted == false
    select new GenericSearchResultModel { Id = table6.Id, Type = GenericModelType.Type4 });
queries.Add(
    from table6 in db.Table6
    where table6.Field1 != null && table6.Field1.StartsWith(search) && table6.Deleted == false
    select new GenericSearchResultModel { Id = table6.Id, Type = GenericModelType.Type4 });
queries.Add(
    from table7 in db.Table7
    where table7.Name.StartsWith(search) && table7.Deleted == false
    select new GenericSearchResultModel { Id = table7.Id, Type = GenericModelType.Type5 });
queries.Add(
    from table8 in db.Table8
    where table8.Field6!.StartsWith(search)
    select new GenericSearchResultModel { Id = table8.Id, Type = GenericModelType.Type6 });

return await queries.Aggregate((a, b) => a.Union(b)).ToListAsync(ctx.CancellationToken);

enum GenericModelType
{
    Type1, Type2, Type3, Type4, Type5, Type6, // ...
}

class GenericSearchResultModel
{
    public long Id { get; set; }
    public GenericModelType Type { get; set; }
}
```

### Environment details

`Linq To DB` version: 5.2.2 (no problems in 5.1.1)

Database (with version): Azure SQL Managed Instance

ADO.NET Provider (with version): Microsoft.Data.SqlClient 5.1.1

Operating system: Ubuntu (production), Windows 10 (dev)

.NET Version: 6.0

**Comment by Shane32** [comment]
I added debug code to tell what inputs the normalizer was receiving and responding with.  Logged are the instance id (`this.GetHashCode()` from within the `UniqueParametersNormalizer` instance), the timestamp (used to determine which query it pertains to), the input and output values. This is what was logged:

|   instance id | timestamp                         | input                     | output                    |
|--------------:|:----------------------------------|:--------------------------|:--------------------------|
|      29483941 | 2023-07-31T09:54:34.9518752-04:00 | id                        | id                        |
|      29483941 | 2023-07-31T09:54:34.9540919-04:00 | firstName                 | firstName                 |
|      29483941 | 2023-07-31T09:54:34.9540997-04:00 | lastName                  | lastName                  |
|      29483941 | 2023-07-31T09:54:34.9541015-04:00 | emailAddress              | emailAddress              |
|      13511990 | 2023-07-31T09:54:36.3439559-04:00 | search                    | search                    |
|      13511990 | 2023-07-31T09:54:36.3454068-04:00 | search                    | search_1                  |
|      13511990 | 2023-07-31T09:54:36.3454264-04:00 | search                    | search_2                  |
|      13511990 | 2023-07-31T09:54:36.3454436-04:00 | search                    | search_3                  |
|      13511990 | 2023-07-31T09:54:36.3454678-04:00 | search                    | search_4                  |
|      13511990 | 2023-07-31T09:54:36.3454814-04:00 | search_4                  | search_4_1                |
|      13511990 | 2023-07-31T09:54:36.3454965-04:00 | search_4                  | search_4_2                |
|      13511990 | 2023-07-31T09:54:36.34573-04:00   | search_4                  | search_4_3                |
|      13511990 | 2023-07-31T09:54:36.345751-04:00  | search_4                  | search_4_4                |
|      13511990 | 2023-07-31T09:54:36.3457633-04:00 | search_4                  | search_4_5                |
|      59207502 | 2023-07-31T09:55:04.1717773-04:00 | search_4                  | search_4                  |
|      59207502 | 2023-07-31T09:55:04.1717918-04:00 | search_4                  | search_4_1                |
|      59207502 | 2023-07-31T09:55:04.1718081-04:00 | search_4                  | search_4_2                |
|      59207502 | 2023-07-31T09:55:04.1718173-04:00 | search_4                  | search_4_3                |
|      59207502 | 2023-07-31T09:55:04.1718338-04:00 | search_4                  | search_4_4                |
|      59207502 | 2023-07-31T09:55:04.171845-04:00  | search_4_4                | search_4_4_1              |
|      59207502 | 2023-07-31T09:55:04.1718544-04:00 | search_4_4                | search_4_4_2              |
|      59207502 | 2023-07-31T09:55:04.1718676-04:00 | search_4_4                | search_4_4_3              |
|      59207502 | 2023-07-31T09:55:04.171877-04:00  | search_4_4                | search_4_4_4              |
|      59207502 | 2023-07-31T09:55:04.1718849-04:00 | search_4_4                | search_4_4_5              |
|       5300652 | 2023-07-31T09:55:31.1712912-04:00 | take                      | take                      |
|      34303972 | 2023-07-31T09:57:09.4228154-04:00 | search_4_4                | search_4_4                |
|      34303972 | 2023-07-31T09:57:09.4228302-04:00 | search_4_4                | search_4_4_1              |
|      34303972 | 2023-07-31T09:57:09.4228399-04:00 | search_4_4                | search_4_4_2              |
|      34303972 | 2023-07-31T09:57:09.4228564-04:00 | search_4_4                | search_4_4_3              |
|      34303972 | 2023-07-31T09:57:09.4228701-04:00 | search_4_4                | search_4_4_4              |
|      34303972 | 2023-07-31T09:57:09.4228853-04:00 | search_4_4_4              | search_4_4_4_1            |
|      34303972 | 2023-07-31T09:57:09.4228948-04:00 | search_4_4_4              | search_4_4_4_2            |
|      34303972 | 2023-07-31T09:57:09.4229066-04:00 | search_4_4_4              | search_4_4_4_3            |
|      34303972 | 2023-07-31T09:57:09.4229183-04:00 | search_4_4_4              | search_4_4_4_4            |
|      34303972 | 2023-07-31T09:57:09.4229253-04:00 | search_4_4_4              | search_4_4_4_5            |

It appears that two things are happening:
1. During the first execution, halfway through processing the query, instead of using `search` as the input, it uses `search_4` as the input.  Note that all 10 variables are used in the query, as shown in the source code above.  So the 10 variables used are `search`, `search_1`, `search_2`, `search_3`, `search_4`, `search_4_1`, `search_4_2`, `search_4_3`, `search_4_4`, and `search_4_5` -- and they can be seen as such in the trace output.
2. During subsequent executions, it acts as through the variable was named `search_4` (and later `search_4_4` and so on).

So I'm wondering:
1. Why doesn't it use `search` as an input to the method for all 10 parameters during the first execution?
2. Why is it running the parameter names through the normalizer at all during a subsequent/cached execution?  Isn't the SQL cached?  Obviously not since in the trace output I can see the longer parameter names.
3. If it is going run the normalizer again, why isn't it using the original parameter name?
4. Does #4229 help in any way here?

My PR #4230 will indirectly fix the issue, but only by the nature that an excessively long parameter name will not throw an exception -- but the underlying bug still exists.

I might be able to write a PR to demonstrate this issue.  I'll work on that next.

**Comment by Shane32** [comment]
Here is a copy of the code required to reproduce the issue:

```cs
[Test]
public async Task TestParameterNormalization([IncludeDataSources(TestProvName.AllSqlServer)] string context)
{
	using var conn = GetDataConnection(context);
	await using var dbTable1 = await conn.CreateTempTableAsync<Table1>("table1");
	await using var dbTable2 = await conn.CreateTempTableAsync<Table2>("table2");
	await using var dbTable3 = await conn.CreateTempTableAsync<Table3>("table3");
	await using var dbTable4 = await conn.CreateTempTableAsync<Table4>("table4");
	await using var dbTable5 = await conn.CreateTempTableAsync<Table5>("table5");
	await using var dbTable6 = await conn.CreateTempTableAsync<Table6>("table6");
	await using var dbTable7 = await conn.CreateTempTableAsync<Table7>("table7");
	await using var dbTable8 = await conn.CreateTempTableAsync<Table8>("table8");

	var query = GenerateQuery("hello");
	_ = await query.ToListAsync();

	var query2 = GenerateQuery("hello");
	_ = await query2.ToListAsync();

	IQueryable<GenericSearchResultModel> GenerateQuery(string search)
	{
		var queries = new List<IQueryable<GenericSearchResultModel>>();
		queries.Add(
			from table1 in dbTable1
			join table2 in dbTable2 on table1.Id equals table2.Table1Id
			where table1.Deleted == false && table2.Field1.StartsWith(search)
			select new GenericSearchResultModel { Id = table1.Id, Type = GenericModelType.Type1 });
		queries.Add(
			from table3 in dbTable3
			where table3.Field2!.StartsWith(search)
			select new GenericSearchResultModel { Id = table3.Id, Type = GenericModelType.Type2 });
		queries.Add(
			from table4 in dbTable4
			where table4.Field2.StartsWith(search)
			select new GenericSearchResultModel { Id = table4.Table3Id, Type = GenericModelType.Type3 });
		queries.Add(
			from table1 in dbTable1
			where table1.Field3.StartsWith(search) && table1.Deleted == false
			select new GenericSearchResultModel { Id = table1.Id, Type = GenericModelType.Type1 });
		queries.Add(
			from table1 in dbTable1
			join table5 in dbTable5 on table1.Id equals table5.Table1Id
			where table5.Field4 == search && table1.Deleted == false
			select new GenericSearchResultModel { Id = table1.Id, Type = GenericModelType.Type1 });
		queries.Add(
			from table1 in dbTable1
			where table1.Field5!.StartsWith(search) && table1.Deleted == false
			select new GenericSearchResultModel { Id = table1.Id, Type = GenericModelType.Type1 });
		queries.Add(
			from table6 in dbTable6
			where table6.Name.StartsWith(search) && table6.Deleted == false
			select new GenericSearchResultModel { Id = table6.Id, Type = GenericModelType.Type4 });
		queries.Add(
			from table6 in dbTable6
			where table6.Field1 != null && table6.Field1.StartsWith(search) && table6.Deleted == false
			select new GenericSearchResultModel { Id = table6.Id, Type = GenericModelType.Type4 });
		queries.Add(
			from table7 in dbTable7
			where table7.Name.StartsWith(search) && table7.Deleted == false
			select new GenericSearchResultModel { Id = table7.Id, Type = GenericModelType.Type5 });
		queries.Add(
			from table8 in dbTable8
			where table8.Field6!.StartsWith(search)
			select new GenericSearchResultModel { Id = table8.Id, Type = GenericModelType.Type6 });

		// union all queries together
		return queries.Aggregate((q1, q2) => q1.Union(q2));
	}
}

public class Table1
{
	public int Id { get; set; }
	public string Field3 { get; set; } = null!;
	public string Field5 { get; set; } = null!;
	public bool Deleted { get; set; }
}

public class Table2
{
	public int Table1Id { get; set; }
	public string Field1 { get; set; } = null!;
}

public class Table3
{
	public int Id { get; set; }
	public string Field2 { get; set; } = null!;
}

public class Table4
{
	public int Table3Id { get; set; }
	public string Field2 { get; set; } = null!;
}

public class Table5
{
	public int Table1Id { get; set; }
	public string Field4 { get; set; } = null!;
}

public class Table6
{
	public int Id { get; set; }
	public string Name { get; set; } = null!;
	public string Field1 { get; set; } = null!;
	public bool Deleted { get; set; }
}

public class Table7
{
	public int Id { get; set; }
	public string Name { get; set; } = null!;
	public bool Deleted { get; set; }
}

public class Table8
{
	public int Id { get; set; }
	public string Field6 { get; set; } = null!;
}

public class GenericSearchResultModel
{
	public int Id { get; set; }
	public GenericModelType Type { get; set; }
}

public enum GenericModelType
{
	Type1,
	Type2,
	Type3,
	Type4,
	Type5,
	Type6
}
```

This code will demonstrate:
1. When the first query executes, it doesn't number the parameter values sequentially.
2. When the second query executes, it doesn't generate the same SQL.

Output:

```sql
-- DataConnection: BeforeExecute
-- SqlServer.2017 (asynchronously)
DECLARE @search NVarChar(4000) -- String
SET     @search = N'hello%'
DECLARE @search_1 NVarChar(4000) -- String
SET     @search_1 = N'hello%'
DECLARE @search_2 NVarChar(4000) -- String
SET     @search_2 = N'hello%'
DECLARE @search_3 NVarChar(4000) -- String
SET     @search_3 = N'hello%'
DECLARE @search_4 NVarChar(4000) -- String
SET     @search_4 = N'hello'
DECLARE @search_4_1 NVarChar(4000) -- String
SET     @search_4_1 = N'hello%'
DECLARE @search_4_2 NVarChar(4000) -- String
SET     @search_4_2 = N'hello%'
DECLARE @search_4_3 NVarChar(4000) -- String
SET     @search_4_3 = N'hello%'
DECLARE @search_4_4 NVarChar(4000) -- String
SET     @search_4_4 = N'hello%'
DECLARE @search_4_5 NVarChar(4000) -- String
SET     @search_4_5 = N'hello%'

SELECT
	[table1_1].[Id],
	0
FROM
	[tempdb]..[#table1] [table1_1]
		INNER JOIN [tempdb]..[#table2] [table2_1] ON [table1_1].[Id] = [table2_1].[Table1Id]
WHERE
	[table1_1].[Deleted] = 0 AND [table2_1].[Field1] LIKE @search ESCAPE N'~'
UNION
SELECT
	[table3_1].[Id],
	1
FROM
	[tempdb]..[#table3] [table3_1]
WHERE
	[table3_1].[Field2] LIKE @search_1 ESCAPE N'~'
UNION
SELECT
	[table4_1].[Table3Id],
	2
FROM
	[tempdb]..[#table4] [table4_1]
WHERE
	[table4_1].[Field2] LIKE @search_2 ESCAPE N'~'
UNION
SELECT
	[table1_2].[Id],
	0
FROM
	[tempdb]..[#table1] [table1_2]
WHERE
	[table1_2].[Field3] LIKE @search_3 ESCAPE N'~' AND
	[table1_2].[Deleted] = 0
UNION
SELECT
	[table1_3].[Id],
	0
FROM
	[tempdb]..[#table1] [table1_3]
		INNER JOIN [tempdb]..[#table5] [table5_1] ON [table1_3].[Id] = [table5_1].[Table1Id]
WHERE
	[table5_1].[Field4] = @search_4 AND [table1_3].[Deleted] = 0
UNION
SELECT
	[table1_4].[Id],
	0
FROM
	[tempdb]..[#table1] [table1_4]
WHERE
	[table1_4].[Field5] LIKE @search_4_1 ESCAPE N'~' AND
	[table1_4].[Deleted] = 0
UNION
SELECT
	[table6_1].[Id],
	3
FROM
	[tempdb]..[#table6] [table6_1]
WHERE
	[table6_1].[Name] LIKE @search_4_2 ESCAPE N'~' AND
	[table6_1].[Deleted] = 0
UNION
SELECT
	[table6_2].[Id],
	3
FROM
	[tempdb]..[#table6] [table6_2]
WHERE
	[table6_2].[Field1] IS NOT NULL AND [table6_2].[Field1] LIKE @search_4_3 ESCAPE N'~' AND
	[table6_2].[Deleted] = 0
UNION
SELECT
	[table7_1].[Id],
	4
FROM
	[tempdb]..[#table7] [table7_1]
WHERE
	[table7_1].[Name] LIKE @search_4_4 ESCAPE N'~' AND
	[table7_1].[Deleted] = 0
UNION
SELECT
	[table8_1].[Id],
	5
FROM
	[tempdb]..[#table8] [table8_1]
WHERE
	[table8_1].[Field6] LIKE @search_4_5 ESCAPE N'~'

-- DataConnection: Query Execution Time (AfterExecute) (async): 00:00:00.0324384

-- DataConnection: Total Execution Time (Completed) (async): 00:00:00.0817919. Rows Count: 0.

-- DataConnection: BeforeExecute
-- SqlServer.2017 (asynchronously)
DECLARE @search_4 NVarChar(4000) -- String
SET     @search_4 = N'hello%'
DECLARE @search_4_1 NVarChar(4000) -- String
SET     @search_4_1 = N'hello%'
DECLARE @search_4_2 NVarChar(4000) -- String
SET     @search_4_2 = N'hello%'
DECLARE @search_4_3 NVarChar(4000) -- String
SET     @search_4_3 = N'hello%'
DECLARE @search_4_4 NVarChar(4000) -- String
SET     @search_4_4 = N'hello'
DECLARE @search_4_4_1 NVarChar(4000) -- String
SET     @search_4_4_1 = N'hello%'
DECLARE @search_4_4_2 NVarChar(4000) -- String
SET     @search_4_4_2 = N'hello%'
DECLARE @search_4_4_3 NVarChar(4000) -- String
SET     @search_4_4_3 = N'hello%'
DECLARE @search_4_4_4 NVarChar(4000) -- String
SET     @search_4_4_4 = N'hello%'
DECLARE @search_4_4_5 NVarChar(4000) -- String
SET     @search_4_4_5 = N'hello%'

SELECT
	[table1_1].[Id],
	0
FROM
	[tempdb]..[#table1] [table1_1]
		INNER JOIN [tempdb]..[#table2] [table2_1] ON [table1_1].[Id] = [table2_1].[Table1Id]
WHERE
	[table1_1].[Deleted] = 0 AND [table2_1].[Field1] LIKE @search_4 ESCAPE N'~'
UNION
SELECT
	[table3_1].[Id],
	1
FROM
	[tempdb]..[#table3] [table3_1]
WHERE
	[table3_1].[Field2] LIKE @search_4_1 ESCAPE N'~'
UNION
SELECT
	[table4_1].[Table3Id],
	2
FROM
	[tempdb]..[#table4] [table4_1]
WHERE
	[table4_1].[Field2] LIKE @search_4_2 ESCAPE N'~'
UNION
SELECT
	[table1_2].[Id],
	0
FROM
	[tempdb]..[#table1] [table1_2]
WHERE
	[table1_2].[Field3] LIKE @search_4_3 ESCAPE N'~' AND
	[table1_2].[Deleted] = 0
UNION
SELECT
	[table1_3].[Id],
	0
FROM
	[tempdb]..[#table1] [table1_3]
		INNER JOIN [tempdb]..[#table5] [table5_1] ON [table1_3].[Id] = [table5_1].[Table1Id]
WHERE
	[table5_1].[Field4] = @search_4_4 AND [table1_3].[Deleted] = 0
UNION
SELECT
	[table1_4].[Id],
	0
FROM
	[tempdb]..[#table1] [table1_4]
WHERE
	[table1_4].[Field5] LIKE @search_4_4_1 ESCAPE N'~' AND
	[table1_4].[Deleted] = 0
UNION
SELECT
	[table6_1].[Id],
	3
FROM
	[tempdb]..[#table6] [table6_1]
WHERE
	[table6_1].[Name] LIKE @search_4_4_2 ESCAPE N'~' AND
	[table6_1].[Deleted] = 0
UNION
SELECT
	[table6_2].[Id],
	3
FROM
	[tempdb]..[#table6] [table6_2]
WHERE
	[table6_2].[Field1] IS NOT NULL AND [table6_2].[Field1] LIKE @search_4_4_3 ESCAPE N'~' AND
	[table6_2].[Deleted] = 0
UNION
SELECT
	[table7_1].[Id],
	4
FROM
	[tempdb]..[#table7] [table7_1]
WHERE
	[table7_1].[Name] LIKE @search_4_4_4 ESCAPE N'~' AND
	[table7_1].[Deleted] = 0
UNION
SELECT
	[table8_1].[Id],
	5
FROM
	[tempdb]..[#table8] [table8_1]
WHERE
	[table8_1].[Field6] LIKE @search_4_4_5 ESCAPE N'~'

-- DataConnection: Query Execution Time (AfterExecute) (async): 00:00:00.0124016

-- DataConnection: Total Execution Time (Completed) (async): 00:00:00.0130714. Rows Count: 0.
```

**Comment by Shane32** [comment]
PR #4229 does not fix the underlying issue here.

**Comment by Shane32** [comment]
This issue will also prevent SQL Server from using its cached execution plan, since the generated SQL changes each time the query is run.

**Comment by Shane32** [comment]
@MaceWindu In order to write a PR that contains the proper test, I want to write a test that uses a DataProvider implementation which has an overridden `GetQueryParameterNormalizer` member.  Is there an easy way to do so?  Then I can provide a mock implementation to validate the inputs to the method.

**Comment by MaceWindu** [comment]
@Shane32 in that case I think it is fine to make it virtual

**Comment by Shane32** [comment]
No what I mean is, what is the right way to write a test like this:
```cs
		private class MyDataProvider : SqlServerDataProvider
		{
			public MyDataProvider() : base(ProviderName.SqlServer2008, SqlServerVersion.v2008, SqlServerProvider.MicrosoftDataSqlClient) { }
			public override IQueryParametersNormalizer GetQueryParameterNormalizer()
			{
				// custom code
			}
		}
		[Test]
		public async Task TestParameterNormalization([IncludeDataSources(TestProvName.AllSqlServer)] string context)
		{
			var options = new DataOptions().UseConfiguration(context);
			options.UseDataProvider(new MyDataProvider());
			using var conn = new TestDataConnection(options);
			//custom code
		}
```
Because in this case it's locked to the SqlServer 2008 dialect with the MicrosoftDataSqlClient provider.  So I just was wondering what would be the proper way to write the test in this situation...as ideally the test should run for many providers.

**Comment by Shane32** [comment]
I'm just not familiar enough with the test framework used here and I was thinking someone probably knows how already.

**Comment by MaceWindu** [comment]
I would do it like that:
```cs
[Test]
public void Test([DataSources(false)] string context)
{
    IDataProvider provider;
    using (var db = CreateDataConnection(context))
    {
        provider = db.Provider;
    }

    // simple proxy clas to delegate all IDataProvider calls to provider instance
    var testProvider = new WrapperProvider(provider);
    ....
}
```

**Comment by Shane32** [comment]
Added failing tests to #4230 although I'm not sure what the next step is to fix the issue.  We could merge #4230 without these additional tests, or revert #4146, or provide additional fixes within #4230
